File: Kernel Simulator/Shell/Shells/RSSShell.vb
﻿
'    Kernel Simulator  Copyright (C) 2018-2022  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Threading
Imports KS.Kernel
Imports KS.Network.RSS.Instance
Imports KS.Network.RSS

Public Class RSSShell
    Inherits ShellExecutor
    Implements IShell

    Public Overrides ReadOnly Property ShellType As ShellType Implements IShell.ShellType
        Get
            Return ShellType.RSSShell
        End Get
    End Property

    Public Overrides Property Bail As Boolean Implements IShell.Bail

    Public Overrides Sub InitializeShell(ParamArray ShellArgs() As Object) Implements IShell.InitializeShell
        'Add handler for RSS shell
        SwitchCancellationHandler(ShellType.RSSShell)
        Dim OldRSSFeedLink As String = ""
        Dim FeedUrl As String = ""
        If ShellArgs.Length > 0 Then
            FeedUrl = ShellArgs(0)
        End If
        RSSFeedLink = FeedUrl

        While Not Bail
Begin:
            If String.IsNullOrWhiteSpace(RSSFeedLink) Then
                Do While String.IsNullOrWhiteSpace(RSSFeedLink)
                    Try
                        If Not String.IsNullOrWhiteSpace(RSSFeedUrlPromptStyle) Then
                            Write(ProbePlaces(RSSFeedUrlPromptStyle), False, ColTypes.Input)
                        Else
                            Write(DoTranslation("Enter an RSS feed URL:") + " ", False, ColTypes.Input)
                        End If
                        RSSFeedLink = Console.ReadLine
                        RSSFeedInstance = New RSSFeed(RSSFeedLink, RSSFeedType.Infer)
                        RSSFeedLink = RSSFeedInstance.FeedUrl
                        OldRSSFeedLink = RSSFeedLink
                    Catch ex As Exception
                        Wdbg(DebugLevel.E, "Failed to parse RSS feed URL {0}: {1}", FeedUrl, ex.Message)
                        WStkTrc(ex)
                        Write(DoTranslation("Failed to parse feed URL:") + " {0}", True, ColTypes.Error, ex.Message)
                        RSSFeedLink = ""
                    End Try
                Loop
            Else
                SyncLock RssShellCancelSync
                    'Make a new RSS feed instance
                    Try
                        If OldRSSFeedLink <> RSSFeedLink Then
                            RSSFeedInstance = New RSSFeed(RSSFeedLink, RSSFeedType.Infer)
                            RSSFeedLink = RSSFeedInstance.FeedUrl
                        End If
                        OldRSSFeedLink = RSSFeedLink
                    Catch ex As Exception
                        Wdbg(DebugLevel.E, "Failed to parse RSS feed URL {0}: {1}", RSSFeedLink, ex.Message)
                        WStkTrc(ex)
                        Write(DoTranslation("Failed to parse feed URL:") + " {0}", True, ColTypes.Error, ex.Message)
                        RSSFeedLink = ""
                        GoTo Begin
                    End Try

                    'Send ping to keep the connection alive
                    If Not RSSKeepAlive And Not RSSRefresher.IsAlive And RSSRefreshFeeds Then RSSRefresher.Start()
                    Wdbg(DebugLevel.I, "Made new thread about RefreshFeeds()")

                    'Prepare for prompt
                    If DefConsoleOut IsNot Nothing Then
                        Console.SetOut(DefConsoleOut)
                    End If
                    Wdbg(DebugLevel.I, "RSSShellPromptStyle = {0}", RSSShellPromptStyle)
                    If RSSShellPromptStyle = "" Then
                        Write("[", False, ColTypes.Gray) : Write("{0}", False, ColTypes.UserName, New Uri(RSSFeedLink).Host) : Write("] > ", False, ColTypes.Gray)
                    Else
                        Dim ParsedPromptStyle As String = ProbePlaces(RSSShellPromptStyle)
                        ParsedPromptStyle.ConvertVTSequences
                        Write(ParsedPromptStyle, False, ColTypes.Gray)
                    End If
                    SetInputColor()

                    'Prompt for command
                    Kernel.KernelEventManager.RaiseRSSShellInitialized(RSSFeedLink)
                    Dim WrittenCommand As String = Console.ReadLine
                    GetLine(WrittenCommand, False, "", ShellType.RSSShell)
                End SyncLock
            End If
        End While

        'Disconnect the session
        If RSSKeepAlive Then
            Wdbg(DebugLevel.W, "Exit requested, but not disconnecting.")
        Else
            Wdbg(DebugLevel.W, "Exit requested. Disconnecting host...")
            If RSSRefreshFeeds Then RSSRefresher.Abort()
            RSSFeedLink = ""
            RSSFeedInstance = Nothing
            RSSRefresher = New Thread(AddressOf RefreshFeeds) With {.Name = "RSS Feed Refresher"}
        End If

        'Remove handler for RSS shell
        SwitchCancellationHandler(LastShellType)
    End Sub

End Class

File: public/Kernel Simulator/Shell/Shells/RSS/RSSShell.cs
﻿
// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using System;
using System.Threading;
using Extensification.StringExts;
using KS.ConsoleBase.Colors;
using KS.ConsoleBase.Inputs;
using KS.Kernel;
using KS.Kernel.Debugging;
using KS.Languages;
using KS.Misc.Probers;
using KS.Misc.Text;
using KS.Misc.Writers.ConsoleWriters;
using KS.Network.RSS;
using KS.Network.RSS.Instance;
using KS.Shell.Prompts;
using KS.Shell.ShellBase.Commands;
using KS.Shell.ShellBase.Shells;

namespace KS.Shell.Shells.RSS
{
    /// <summary>
    /// The RSS shell
    /// </summary>
    public class RSSShell : BaseShell, IShell
    {

        /// <inheritdoc/>
        public override string ShellType => "RSSShell";

        /// <inheritdoc/>
        public override bool Bail { get; set; }

        /// <inheritdoc/>
        public override void InitializeShell(params object[] ShellArgs)
        {
            // Handle the RSS feed link provided by user
            bool BailFromEnter = false;
            string OldRSSFeedLink = "";
            string FeedUrl = "";
            if (ShellArgs.Length > 0)
            {
                FeedUrl = Convert.ToString(ShellArgs[0]);
            }
            RSSShellCommon.RSSFeedLink = FeedUrl;

            while (!BailFromEnter)
            {
                if (string.IsNullOrWhiteSpace(RSSShellCommon.RSSFeedLink))
                {
                    while (string.IsNullOrWhiteSpace(RSSShellCommon.RSSFeedLink))
                    {
                        try
                        {
                            if (!string.IsNullOrWhiteSpace(RSSShellCommon.RSSFeedUrlPromptStyle))
                            {
                                TextWriterColor.Write(PlaceParse.ProbePlaces(RSSShellCommon.RSSFeedUrlPromptStyle), false, ColorTools.ColTypes.Input);
                            }
                            else
                            {
                                TextWriterColor.Write(Translate.DoTranslation("Enter an RSS feed URL:") + " ", false, ColorTools.ColTypes.Input);
                            }
                            RSSShellCommon.RSSFeedLink = Input.ReadLine();
                            if (ReadLineReboot.ReadLine.ReadRanToCompletion)
                            {
                                // The user entered the feed URL
                                RSSShellCommon.RSSFeedInstance = new RSSFeed(RSSShellCommon.RSSFeedLink, RSSFeedType.Infer);
                                RSSShellCommon.RSSFeedLink = RSSShellCommon.RSSFeedInstance.FeedUrl;
                                OldRSSFeedLink = RSSShellCommon.RSSFeedLink;
                            }
                            else
                            {
                                // We're aborting.
                                Bail = true;
                            }
                            BailFromEnter = true;
                        }
                        catch (ThreadInterruptedException)
                        {
                            Flags.CancelRequested = false;
                            BailFromEnter = true;
                            Bail = true;
                        }
                        catch (Exception ex)
                        {
                            DebugWriter.WriteDebug(DebugLevel.E, "Failed to parse RSS feed URL {0}: {1}", FeedUrl, ex.Message);
                            DebugWriter.WriteDebugStackTrace(ex);
                            TextWriterColor.Write(Translate.DoTranslation("Failed to parse feed URL:") + " {0}", true, ColorTools.ColTypes.Error, ex.Message);
                            RSSShellCommon.RSSFeedLink = "";
                        }
                    }
                }
                else
                {
                    // Make a new RSS feed instance
                    try
                    {
                        if ((OldRSSFeedLink ?? "") != (RSSShellCommon.RSSFeedLink ?? ""))
                        {
                            if (RSSShellCommon.RSSFeedLink == "select")
                            {
                                RSSTools.OpenFeedSelector();
                            }
                            RSSShellCommon.RSSFeedInstance = new RSSFeed(RSSShellCommon.RSSFeedLink, RSSFeedType.Infer);
                            RSSShellCommon.RSSFeedLink = RSSShellCommon.RSSFeedInstance.FeedUrl;
                        }
                        OldRSSFeedLink = RSSShellCommon.RSSFeedLink;
                        BailFromEnter = true;
                    }
                    catch (ThreadInterruptedException)
                    {
                        Flags.CancelRequested = false;
                        BailFromEnter = true;
                        Bail = true;
                    }
                    catch (Exception ex)
                    {
                        DebugWriter.WriteDebug(DebugLevel.E, "Failed to parse RSS feed URL {0}: {1}", RSSShellCommon.RSSFeedLink, ex.Message);
                        DebugWriter.WriteDebugStackTrace(ex);
                        TextWriterColor.Write(Translate.DoTranslation("Failed to parse feed URL:") + " {0}", true, ColorTools.ColTypes.Error, ex.Message);
                        RSSShellCommon.RSSFeedLink = "";
                    }
                }
            }

            while (!Bail)
            {
                try
                {
                    // Send ping to keep the connection alive
                    if (!RSSShellCommon.RSSKeepAlive & !RSSShellCommon.RSSRefresher.IsAlive & RSSShellCommon.RSSRefreshFeeds)
                        RSSShellCommon.RSSRefresher.Start();
                    DebugWriter.WriteDebug(DebugLevel.I, "Made new thread about RefreshFeeds()");

                    // Prompt for the command
                    Shell.GetLine();
                }
                catch (ThreadInterruptedException)
                {
                    Flags.CancelRequested = false;
                    Bail = true;
                }
                catch (Exception ex)
                {
                    DebugWriter.WriteDebugStackTrace(ex);
                    TextWriterColor.Write(Translate.DoTranslation("There was an error in the shell.") + CharManager.NewLine + "Error {0}: {1}", true, ColorTools.ColTypes.Error, ex.GetType().FullName, ex.Message);
                    continue;
                }
            }

            // Disconnect the session
            if (RSSShellCommon.RSSKeepAlive)
            {
                DebugWriter.WriteDebug(DebugLevel.W, "Exit requested, but not disconnecting.");
            }
            else
            {
                DebugWriter.WriteDebug(DebugLevel.W, "Exit requested. Disconnecting host...");
                if (RSSShellCommon.RSSRefreshFeeds)
                    RSSShellCommon.RSSRefresher.Stop();
                RSSShellCommon.RSSFeedLink = "";
                RSSShellCommon.RSSFeedInstance = null;
            }
        }

    }
}
